pkg/resources/ali/ali.go

```diff
@@ -0,0 +1,691 @@
+}
+
+func DeleteLoadBalancer(cloud fi.Cloud, r *resources.Resource) error {
+	c := cloud.(aliup.ALICloud)
+	glog.V(2).Infof("Removing LoadBalancer with Id %s", r.ID)
+
+	err := c.SlbClient().DeleteLoadBalancer(r.ID)
+	if err != nil {
+		return fmt.Errorf("err deleting LoadBalancer:%v", err)
+	}
+	return nil
+}
+
+func (d *clusterDiscoveryALI) ListSecurityGroup() ([]*resources.Resource, error) {
+	var resourceTrackers []*resources.Resource
+	// SecurityGroup's name can be renamed in alicloud.
+	// We recognize SecurityGroup with its name and clusterTags
+	tags := make(map[string]string)
+	d.aliCloud.AddClusterTags(tags)
+
+	securityGroupNames := []string{
+		"masters." + d.clusterName,
+		"bastions." + d.clusterName,
+		"nodes." + d.clusterName,
+	}
+
+	describeResourceByTagsArgs := &ecs.DescribeResourceByTagsArgs{
+		ResourceType: ecs.TagResourceType(alitasks.SecurityResource),
+		RegionId:     common.Region(d.aliCloud.Region()),
+		Tag:          tags,
+	}
+
+	resourceList, _, err := d.aliCloud.EcsClient().DescribeResourceByTags(describeResourceByTagsArgs)
+	if err != nil {
+		return nil, fmt.Errorf("err listing securityGroup:%v", err)
+	}
+
+	blocked := []string{}
+	groupTrackers := []*resources.Resource{}
+
+	if len(resourceList) != 0 {
+		for _, resource := range resourceList {
+			find := false
+
+			describeSecurityGroupAttributeArgs := &ecs.DescribeSecurityGroupAttributeArgs{
+				SecurityGroupId: resource.ResourceId,
+				RegionId:        resource.RegionId,
+			}
+			respGroup, err := d.aliCloud.EcsClient().DescribeSecurityGroupAttribute(describeSecurityGroupAttributeArgs)
+			if err != nil {
+				return nil, fmt.Errorf("err listing securityGroup:%v", err)
+			}
+
+			for _, value := range securityGroupNames {
+				if respGroup.SecurityGroupName == value {
+					find = true
+				}
+			}
+
+			if find {
+				groupTracker := &resources.Resource{
+					Name:    respGroup.SecurityGroupName,
+					ID:      respGroup.SecurityGroupId,
+					Type:    typeSecurityGroup,
+					Deleter: DeleteSecurityGroup,
+				}
+				groupTrackers = append(groupTrackers, groupTracker)
+
+				if len(respGroup.Permissions.Permission) != 0 {
+					for i, securityGroupRole := range respGroup.Permissions.Permission {
+						roleTracker := &resources.Resource{
+							Name:    respGroup.SecurityGroupId,
+							ID:      respGroup.SecurityGroupName + strconv.Itoa(i),
+							Type:    typeSecurityGroupRole,
+							Deleter: DeleteSecurityGroupRole,
+							Obj:     securityGroupRole,
+						}
+						resourceTrackers = append(resourceTrackers, roleTracker)
+
+						// Any SecurityGroupRole which depends on the current SecurityGroup should be deleted before delete the SecurityGroup.
+						// So before deleting any SecurityGroup, we should delete all SecurityGroupRoles of all SecurityGroups.
+						blocked = append(blocked, roleTracker.Type+":"+roleTracker.ID)
+					}
+				}
+
+			}
+		}
+	}
+
+	if len(groupTrackers) != 0 {
+		for _, groupTracker := range groupTrackers {
```

Comment:
FYI you don't have to do the `if len(groupTrackers) != 0` check here - we do it when we're making an API call with the list of items, and we sometimes do it for clarity (to put the code into a block).  But here you could just drop it and loop over groupTrackers (you can loop over a nil slice, it's treated the same as an empty slice)